### ISSUE TITLE
Pattern.record appears to be broken

### ISSUE THREAD

**Issue by jamiehodge** [supercollider/supercollider#5879]
The inlining of `pattern.record` (https://github.com/supercollider/supercollider/pull/5275) is resulting in header-only files. I copied the previous, server recorder-based method into the class definition and it created a valid audio recording.

## Environment

* SuperCollider version: 3.12.2
* Operating system: Mac OS
* Other details (Qt version, audio driver, etc.):

## Steps to reproduce

```supercollider
pattern.record(dur: 10)
```

## Expected vs. actual behavior

AIFF file with 100 or so bytes of headers and no recording.

**Comment by jamshark70** [comment]
```
(
Pbind(
	\degree, Pn(Pseries(0, 1, 8), 2),
	\dur, 0.25
).record("~/p-rec-test.wav".standardizePath, "wav", "int16", 2);
)
```

![sc-pat-rec-01](https://user-images.githubusercontent.com/318301/192912501-7a6db1e3-ca02-4e8f-a341-307b6d200b0c.png)

So the inlining is not broken. The recording logic is fine. (Actually the inlining was done for a reason -- you already pointed to the issue; it would be worth reading the history to understand why it was changed.)

What *does* appear to be a problem is that there is no default header or sample format.

So the fix would be:

```
	record { |path, headerFormat, sampleFormat, numChannels = 2, dur = nil, fadeTime = 0.2, clock(TempoClock.default), protoEvent(Event.default), server(Server.default), out = 0, outNumChannels|

		var buffer, nodeID, defname;
		var pattern = if(dur.notNil) { Pfindur(dur, this) } { this };

		if(headerFormat.isNil) {
			headerFormat = server.recHeaderFormat;
		};
		if(sampleFormat.isNil) {
			sampleFormat = server.recSampleFormat;
		};

		server.waitForBoot {
... etc etc...
		}
	}
```

**Comment by telephon** [comment]
Btw. was there a reason that it wasn't possible to use the built in `Recorder` class for this?

**Comment by jamshark70** [comment]
> was there a reason that it wasn't possible to use the built in Recorder class for this?

This information is not hard to find in the publicly available history. The link is right there at the top of the issue text...

https://github.com/supercollider/supercollider/pull/5275: "https://scsynth.org/t/calling-supercollider-script-from-windows-batch-file/2964/8 reports that an audio file produced by pattern.record begins with some silence, where a reasonable expectation would be that the recording begins exactly at the onset of the pattern's audio."

https://scsynth.org/t/calling-supercollider-script-from-windows-batch-file/2964/8:

1. (Original poster) "On a slightly different topic, I get about 100ms of silence at the beginning of each recording."

2. (Me) "Recording requires some initialization, and this is asynchronous. It allocates some memory from the OS. There is no way to predict how long that will take. So, it has to be outside of the audio thread, and this means that a record convenience method cannot guarantee that recording will start precisely in sync with audio. The other thing: Patterns play with s.latency delay by default. (See the Server Timing help file.) But it takes some time to initialize, so I guess you’re seeing server latency minus recording setup time."

So the correct way to record a pattern is to set up the buffer and b_write first, then `server.sync`, and then start the recording synth with the same timestamp as the first synth that the pattern produces. I cannot imagine any way that Recorder could do this = it is wrong to rely on Recorder for this. (Recorder is a convenience class. There is never any guarantee that a convenience class will cover all possible cases.)

(Also, as a side note, as I hinted on the forum, I am feeling some strain lately from ongoing back-and-forth on questions that don't really need back-and-forth. I had just extricated myself from a couple of those on the forum, and wake up this morning to find this question kicked back to me [when the answer is already publicly available], and another one on the forum. So as of now, I am going to temporarily unsubscribe from this repository.)
